test/distributed/fsdp/test_fsdp_clip_grad_norm.py

```diff
@@ -0,0 +1,136 @@
+# Owner(s): ["oncall: distributed"]
+
+import sys
+from math import inf
+
+import torch
+from torch import distributed as dist
+from torch.distributed.fsdp.fully_sharded_data_parallel import (
+    FullyShardedDataParallel as FSDP,
+    CPUOffload,
+    _calc_grad_norm,
+)
+from torch.nn import utils as nn_utils
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import (
+    DeterministicModel,
+    FSDPTest,
+)
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    run_tests,
+    parametrize,
+    instantiate_parametrized_tests,
+)
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+def _collect_model_total_norms(model, norm_type):
+    if norm_type == inf:
+        return max(p.grad.abs().max() for p in model.parameters())
+    else:
+        total_norm = 0.0
+        for p in model.parameters():
+            local_norm = torch.linalg.norm(p.grad, norm_type, dtype=torch.float32)
+            total_norm += local_norm ** norm_type
+        return total_norm ** (1.0 / norm_type)
+
+
+class TestClipGradNorm(FSDPTest):
```

Comment:
Let's add a test for nested FSDP model, then overall it looks great!

Reply:
Sure.